File: Data Structures & Algorithms/two-integer-sum/submission-5.cpp
class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        std::unordered_map<int, int> seen;

        for (int i=0; i < nums.size(); ++i) {
            const int diff = target - nums[i];
            auto it = seen.find(diff);
            if (it != seen.end()) {
                return {it->second,i};
            }
            else {
                seen.insert({nums[i],i});
            }
        }
        return vector<int>();
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-5.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        out = [1] * n
        for i in range(1,n):
            out[i] = out[i-1] * nums[i-1]

        running_suffix = 1
        for j in range(n-1,-1,-1):
            out[j] *= running_suffix
            running_suffix *= nums[j]
        
        return out